### PR TITLE
feat: create Production plan from sales order

### PR DESCRIPTION
**Pull request by nmimsnikhil**
[Asana](https://app.asana.com/0/1192407899754728/1199140629720845)
Create production plan from sales order dashboard.
Moved function from bloomstack_core to erpnext.
https://github.com/Bloomstack/erpnext/pull/771

pp gif

multiple pp

### CODE REVIEW

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 1]
```diff
@@ -1071,3 +1071,196 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+			"field_map": {
+				"item_code": "item_code",
+				'name': 'sales_order_item',
+				"qty": "planned_qty",
+				"parent": "sales_order"
+			},
+			'postprocess': update_item_data
+		}
+	}, target_doc)
+	return doc
```
Comment (nmimsnikhil) [inline erpnext/selling/doctype/sales_order/sales_order.py]: create Production plan and assign values to the items.

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 2]
```diff
@@ -1071,3 +1071,196 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+
+		created_orders.append({
+			"sales_order": order,
+			"customer": customer,
+			"delivery_notes": delivery_notes,
+			"created": created
+		})
+
+	return created_orders
+
```
Comment (nmimsnikhil) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Moved from bloomstack_core.

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 3]
```diff
@@ -1071,3 +1071,196 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+			production_plans = []
+
+		created_orders.append({
+			"sales_order": order,
+			"customer": customer,
+			"production_plans": production_plans,
+			"created": created
+		})
+
+	return created_orders
```
Comment (nmimsnikhil) [inline erpnext/selling/doctype/sales_order/sales_order.py]: To create different production plan of multiples sales order.

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 4]
```diff
@@ -54,5 +54,341 @@ frappe.listview_settings['Sales Order'] = {
...
+										delete r._server_messages;
+									}
+
+									listview.refresh();
+								}
+							}
+						});
+					})
+			};
+		};
```
Comment (nmimsnikhil) [inline erpnext/selling/doctype/sales_order/sales_order_list.js]: Moved from bloomstack_core.

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 5]
```diff
@@ -54,5 +54,341 @@ frappe.listview_settings['Sales Order'] = {
...
+									message: res.message,
+									doc: {}
+								})
+							}
+						})
+					}
+				})
+			}
+		}
+		listview.page.add_actions_menu_item(__('Email'), send_email_action, true);
```
Comment (nmimsnikhil) [inline erpnext/selling/doctype/sales_order/sales_order_list.js]: Moved from bloomstack_core

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 6]
```diff
@@ -1071,3 +1071,196 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
 
 	frappe.db.set_value('Sales Order Item', sales_order_item, 'produced_qty', total_produced_qty)
+
+"""
+Creating Production Plan from Sales Order
+Args:
+source_name: string -> name of Sales Order through which Production Plan will form.
+Return:
+dictionary -> it contains the all the value assigned to the production plan.
+"""
```
Comment (vishdha) [inline erpnext/selling/doctype/sales_order/sales_order.py]: @nmimsnikhil it should be inside the function not outside

**Review by vishdha — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 7]
```diff
@@ -1071,3 +1071,196 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+
+	return created_orders
+
+"""
+Creating Multiple Production Plan from Sales Order
+Args:
+order: list -> it contains list of name of sales_order.
+Return:
+list -> list of created production_plan along with their sales_order_name .
+"""
```
Comment (vishdha) [inline erpnext/selling/doctype/sales_order/sales_order.py]: it should be inside the function

**Review by vishdha — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 8]
```diff
@@ -54,5 +54,341 @@ frappe.listview_settings['Sales Order'] = {
...
+				for (let doc of selected_docs) {
+					if (doc.docstatus !== 1) {
+						frappe.throw(__("Cannot Email Draft or cancelled documents"));
+					}
+					if (doc.title !== title) {
+						frappe.throw(__("Select only one customer's sales orders"));
+					}
+				}
+				frappe.call({
+					method: "bloomstack_core.utils.get_contact",
```
Comment (vishdha) [inline erpnext/selling/doctype/sales_order/sales_order_list.js]: @nmimsnikhil here is a dependency with bloomstack module. move this stuff to erpnext

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 9]
```diff
@@ -54,5 +54,341 @@ frappe.listview_settings['Sales Order'] = {
...
+					args: { "doctype": doctype, "name": selected_docs[0].name, "contact_field": "customer" },
+					callback: function (r) {
+						frappe.call({
+							method: "bloomstack_core.utils.get_document_links",
+							args: { "doctype": doctype, "docs": selected_docs },
```
Comment (vishdha) [inline erpnext/selling/doctype/sales_order/sales_order_list.js:378]: @nmimsnikhil here is a dependency with bloomstack module. move this stuff to erpnext.

**Review by vishdha — CHANGES_REQUESTED** [review]

**Review by rvpasari — CHANGES_REQUESTED** [review]
All these functions are elaborate, and as specified by Himanshu in the group, will require comments and docstrings. Ideally, a dev should not have to enter the function definition and read through what the function is doing and what value (and type of value) its returning. All of it can be put in the docstring

**Review by rvpasari — CHANGES_REQUESTED** [review]
Please update all the functions marked in green with docstrings and comments. (When something shows up highlighted in green, it automatically means that the code snippet has been added/modified by the dev).

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 10]
```diff
@@ -1071,3 +1071,239 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+	Args:
+		source_name: string -> name of Sales Order through which Production Plan will form
+	Return:
+		dictionary -> it contains the all the value assigned to the production plan"""
+
+	# Set values to production plan fetch from sales order
+	def set_missing_values(source, target):
+		target.get_items_from = source.doctype
+		target.append("sales_orders", {
+			"sales_order": "SAL-ORD-2020-00293",
```
Comment (vishdha) [inline erpnext/selling/doctype/sales_order/sales_order.py]: @nmimsnikhil Remove this and make the appropriate change here

**Review by vishdha — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/sales_order/sales_order_list.js [hunk 11]
```diff
@@ -54,5 +54,341 @@ frappe.listview_settings['Sales Order'] = {
...
+					if (doc.title !== title) {
+						frappe.throw(__("Select only one customer's sales orders"));
+					}
+				}
+				frappe.call({
+					method: "erpnext.utils.get_contact",
+					args: { "doctype": doctype, "name": selected_docs[0].name, "contact_field": "customer" },
+					callback: function (r) {
+						frappe.call({
+							method: "erpnext.utils.get_document_links",
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order_list.js:377]: Check the frappe.call method

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 12]
```diff
@@ -1071,3 +1071,210 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+			"customer": source.customer,
+			"grand_total": source.base_grand_total
+		})
+
+	# Updates the fields of items.
+	def update_item_data(source, target, source_parent):
+		bom_no = frappe.get_value("BOM", {
+			"item": target.item_code,
+			"is_default": 1
+			})
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
		bom_no = frappe.get_value("BOM", {"item": target.item_code, "is_default": 1})
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 13]
```diff
@@ -1071,3 +1071,210 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+		if bom_no:
+			bom_data = frappe.db.get_value('BOM', bom_no, ['raw_material_cost','operating_cost'], as_dict=1)
+			target.bom_no = bom_no
+			target.raw_material_cost = bom_data.raw_material_cost
+			target.total_operational_cost = bom_data.total_operating_cost
+			target.total_operational_hours = flt(frappe.db.get_value("BOM Operation", {"parent": bom_no}, "sum(time_in_mins)")) / 60.0
+			target.total_workstations = frappe.db.count("BOM Operation", {"parent": bom_no, "workstation": ["!=", ""]})
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
			target.total_workstations = frappe.db.count("BOM Operation", {"parent": bom_no, "workstation": ["is", "set"]})
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 14]
```diff
@@ -1071,3 +1071,210 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+		created = False
+		customer = frappe.db.get_value("Sales Order", order, "customer")
+
+		# check if a Pick List already exists against the order.
+		pick_lists = frappe.get_all("Pick List",
+			filters=[
+				["Pick List", "docstatus", "<", 2],
+				["Pick List Item", "sales_order", "=", order]
+			],
+			distinct=True)
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
		pick_lists = frappe.get_all("Pick List", filters=[
				["Pick List", "docstatus", "<", 2],
				["Pick List Item", "sales_order", "=", order]
			], distinct=True)
```

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 15]
```diff
@@ -1071,3 +1071,210 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+		pick_lists = [item.name for item in pick_lists if item.name]
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]:
```suggestion
```
You dont have to loop over it again. ORM will always return data. Use `as_list` param for get_all

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 16]
```diff
@@ -1071,3 +1071,210 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+		customer = frappe.db.get_value("Sales Order", order, "customer")
+
+		# check if a Sales Invoice already exists against the order.
+		sales_invoices = frappe.get_all("Sales Invoice",
+			filters=[
+				["Sales Invoice", "docstatus", "<", 2],
+				["Sales Invoice Item", "sales_order", "=", order]
+			],
+			distinct=True)
+		sales_invoices = [item.name for item in sales_invoices if item.name]
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Similar to the above change

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 17]
```diff
@@ -1071,3 +1071,210 @@ def update_produced_qty_in_so_item(sales_order, sales_order_item):
...
+		customer = frappe.db.get_value("Sales Order", order, "customer")
+
+		# check if a Delivery Note already exists against the order.
+		delivery_notes = frappe.get_all("Delivery Note",
+			filters=[
+				["Delivery Note", "docstatus", "<", 2],
+				["Delivery Note Item", "against_sales_order", "=", order]
+			],
+			distinct=True)
+		delivery_notes = [item.name for item in delivery_notes if item.name]
```
Comment (hrwX) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Same changes as suggested above

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — APPROVED** [review]

**Review by vishdha — APPROVED** [review]

**Pull request merged**